includes/payment-methods/class-upe-payment-gateway.php

```diff
@@ -795,13 +802,19 @@ public function process_redirect_payment( $order_id, $intent_id, $save_payment_m
 		} catch ( Exception $e ) {
 			Logger::log( 'Error: ' . $e->getMessage() );
 
-			// Confirm our needed variables are set before using them due to there could be a server issue during the get_intent process.
-			$status    = $status ?? null;
-			$charge_id = $charge_id ?? null;
+			$is_order_id_mismatched_exception =
+				is_a( $e, Process_Payment_Exception::class )
+				&& 'upe_process_redirect_order_id_mismatched' === $e->get_error_code();
 
-			/* translators: localized exception message */
-			$message = sprintf( __( 'UPE payment failed: %s', 'woocommerce-payments' ), $e->getMessage() );
-			$this->order_service->mark_payment_failed( $order, $intent_id, $status, $charge_id, $message );
+			if ( ! $is_order_id_mismatched_exception ) {
+				// Confirm our needed variables are set before using them due to there could be a server issue during the get_intent process.
+				$status    = $status ?? null;
+				$charge_id = $charge_id ?? null;
+
+				/* translators: localized exception message */
+				$message = sprintf( __( 'UPE payment failed: %s', 'woocommerce-payments' ), $e->getMessage() );
+				$this->order_service->mark_payment_failed( $order, $intent_id, $status, $charge_id, $message );
```

Comment:
Any reason to skip this? A meaningful note could provide context here, and AFAICT there are no downsides to calling `mark_payment_failed()`.

Was there a specific reason this was added?

Reply:
I've added this commit to explain why that's the case 824dad021cef1716b53172af3225ec4729c72e14

Basically, when we detect the order ID mismatch here, the provided intent ID is no longer relevant for the current processing order.